Install fails on Linux (archlinux, gcc 5.3.0, glibc 2.22, ruby-2.3.0)

Unfortunately I cannot install nmatrix on my 64bit linux box. The error message looks like this:

```
~/src/external/nmatrix bundle exec rake compile
cd tmp/x86_64-linux/nmatrix/2.3.0
make
compiling ../../../../ext/nmatrix/nmatrix.cpp
In file included from ../../../../ext/nmatrix/nmatrix.cpp:283:0:
../../../../ext/nmatrix/ruby_nmatrix.c: In function ‘VALUE nm_reshape_bang(VALUE, VALUE)’:
../../../../ext/nmatrix/ruby_nmatrix.c:1102:11: warning: unused variable ‘elem’ [-Wunused-variable]
     void* elem = s->elements;
           ^
In file included from /usr/include/ruby-2.3.0/ruby.h:33:0,
                 from ../../../../ext/nmatrix/nmatrix.cpp:34:
../../../../ext/nmatrix/ruby_nmatrix.c: In function ‘nm::dtype_t nm_dtype_guess(VALUE)’:
/usr/include/ruby-2.3.0/ruby/ruby.h:882:52: error: ‘__builtin_choose_expr’ was not declared in this scope
      OBJ_WB_UNPROTECT((VALUE)(obj)), ((VALUE)(obj))))
                                                    ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1039:48: note: in definition of macro ‘RARRAY_CONST_PTR’
 #define RARRAY_CONST_PTR(a) rb_array_const_ptr(a)
                                                ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1059:50: note: in expansion of macro ‘RB_OBJ_WB_UNPROTECT_FOR’
 #define RARRAY_PTR(a) ((VALUE *)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
                                                  ^
../../../../ext/nmatrix/ruby_nmatrix.c:2639:27: note: in expansion of macro ‘RARRAY_PTR’
     return nm_dtype_guess(RARRAY_PTR(v)[0]);
                           ^
../../../../ext/nmatrix/ruby_nmatrix.c: In function ‘void* interpret_initial_value(VALUE, nm::dtype_t)’:
/usr/include/ruby-2.3.0/ruby/ruby.h:882:52: error: ‘__builtin_choose_expr’ was not declared in this scope
      OBJ_WB_UNPROTECT((VALUE)(obj)), ((VALUE)(obj))))
                                                    ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1039:48: note: in definition of macro ‘RARRAY_CONST_PTR’
 #define RARRAY_CONST_PTR(a) rb_array_const_ptr(a)
                                                ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1059:50: note: in expansion of macro ‘RB_OBJ_WB_UNPROTECT_FOR’
 #define RARRAY_PTR(a) ((VALUE *)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
                                                  ^
../../../../ext/nmatrix/ruby_nmatrix.c:2801:23: note: in expansion of macro ‘RARRAY_PTR’
       rubyval_to_cval(RARRAY_PTR(arg)[index], dtype, (char*)init_val + (index * DTYPE_SIZES[dtype]));
                       ^
../../../../ext/nmatrix/ruby_nmatrix.c: In function ‘size_t* interpret_shape(VALUE, size_t*)’:
/usr/include/ruby-2.3.0/ruby/ruby.h:882:52: error: ‘__builtin_choose_expr’ was not declared in this scope
      OBJ_WB_UNPROTECT((VALUE)(obj)), ((VALUE)(obj))))
                                                    ^
/usr/include/ruby-2.3.0/ruby/ruby.h:691:51: note: in definition of macro ‘RB_FIX2UINT’
 #define RB_FIX2UINT(x) ((unsigned int)rb_fix2uint(x))
                                                   ^
../../../../ext/nmatrix/ruby_nmatrix.c:2828:22: note: in expansion of macro ‘FIX2UINT’
       shape[index] = FIX2UINT( RARRAY_PTR(arg)[index] );
                      ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1059:33: note: in expansion of macro ‘RARRAY_CONST_PTR’
 #define RARRAY_PTR(a) ((VALUE *)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
                                 ^
/usr/include/ruby-2.3.0/ruby/ruby.h:1059:50: note: in expansion of macro ‘RB_OBJ_WB_UNPROTECT_FOR’
 #define RARRAY_PTR(a) ((VALUE *)RARRAY_CONST_PTR(RB_OBJ_WB_UNPROTECT_FOR(ARRAY, a)))
                                                  ^
../../../../ext/nmatrix/ruby_nmatrix.c:2828:32: note: in expansion of macro ‘RARRAY_PTR’
       shape[index] = FIX2UINT( RARRAY_PTR(arg)[index] );
                                ^
Makefile:239: recipe for target 'nmatrix.o' failed
make: *** [nmatrix.o] Error 1
rake aborted!
Command failed with status (2): [make...]
/home/ram/.gem/ruby/2.3.0/gems/rake-compiler-0.9.5/lib/rake/extensiontask.rb:155:in `block (2 levels) in define_compile_tasks'
/home/ram/.gem/ruby/2.3.0/gems/rake-compiler-0.9.5/lib/rake/extensiontask.rb:154:in `block in define_compile_tasks'
Tasks: TOP => compile => compile:x86_64-linux => compile:nmatrix:x86_64-linux => copy:nmatrix:x86_64-linux:2.3.0 => tmp/x86_64-linux/nmatrix/2.3.0/nmatrix.so
(See full trace by running task with --trace)
```

I m also getting these same errors!! What should i Do to remove these errors? 

compiling ../../../../ext/nmatrix/nmatrix.cpp
In file included from ../../../../ext/nmatrix/nmatrix.cpp:283:0:
../../../../ext/nmatrix/ruby_nmatrix.c: In function ‘VALUE nm_reshape_bang(VALUE, VALUE)’:
../../../../ext/nmatrix/ruby_nmatrix.c:1102:11: warning: unused variable ‘elem’ [-Wunused-variable]
     void* elem = s->elements;
           ^
../../../../ext/nmatrix/ruby_nmatrix.c: In function ‘VALUE nm_mset(int, VALUE*, VALUE)’:
../../../../ext/nmatrix/ruby_nmatrix.c:2013:109: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 4 has type ‘size_t {aka unsigned int}’ [-Wformat=]
     rb_raise(rb_eArgError, "wrong number of arguments (%d for %lu)", argc, effective_dim(NM_STORAGE(self))+1);
                                                                                                             ^
../../../../ext/nmatrix/ruby_nmatrix.c: In function ‘VALUE nm_xslice(int, VALUE*, void* (*)(const STORAGE*, SLICE*), void (*)(NMATRIX*), VALUE)’:
../../../../ext/nmatrix/ruby_nmatrix.c:2251:92: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 4 has type ‘size_t {aka unsigned int}’ [-Wformat=]
     rb_raise(rb_eArgError, "wrong number of arguments (%d for %lu)", argc, effective_dim(s));
                                                                                            ^
../../../../ext/nmatrix/ruby_nmatrix.c: In function ‘SLICE* get_slice(size_t, int, VALUE*, size_t*)’:
../../../../ext/nmatrix/ruby_nmatrix.c:2729:106: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘size_t {aka unsigned int}’ [-Wformat=]
       rb_raise(rb_eRangeError, "slice is larger than matrix in dimension %lu (slice component %lu)", r, t);
                                                                                                          ^
../../../../ext/nmatrix/ruby_nmatrix.c:2729:106: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 4 has type ‘size_t {aka unsigned int}’ [-Wformat=]
In file included from /home/rounak/.rvm/rubies/ruby-2.2.1/include/ruby-2.2.0/ruby.h:33:0,
                 from ../../../../ext/nmatrix/nmatrix.cpp:34:
../../../../ext/nmatrix/ruby_nmatrix.c: In function ‘void* interpret_initial_value(VALUE, nm::dtype_t)’:
/home/rounak/.rvm/rubies/ruby-2.2.1/include/ruby-2.2.0/ruby/ruby.h:902:28: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
      RARRAY(a)->as.heap.len)
                            ^
../../../../ext/nmatrix/ruby_nmatrix.c:2800:29: note: in expansion of macro ‘RARRAY_LEN’
     for (index = 0; index < RARRAY_LEN(arg); ++index) {
                             ^
In file included from ../../../../ext/nmatrix/nmatrix.cpp:283:0:
../../../../ext/nmatrix/ruby_nmatrix.c: In function ‘VALUE nm_init_new_version(int, VALUE*, VALUE)’:
../../../../ext/nmatrix/ruby_nmatrix.c:1299:60: warning: ‘init’ may be used uninitialized in this function [-Wmaybe-uninitialized]
     nm_unregister_values(reinterpret_cast<VALUE*>(init), 1);
                                                            ^

Along with these errors , I m also getting many warnings!! I know warnings won't affect the code but still :( 

compiling ../../../../ext/nmatrix/storage/yale/yale.cpp
../../../../ext/nmatrix/storage/yale/yale.cpp: In function ‘void nm_init_yale_functions()’:
../../../../ext/nmatrix/storage/yale/yale.cpp:1157:136: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘unsigned int’ [-Wformat=]
     rb_raise(rb_eStandardError, "unhandled length for sizeof(IType): %lu; note that IType is probably defined as size_t", sizeof(IType));
                                                                                                                                        ^
../../../../ext/nmatrix/storage/yale/yale.cpp: In function ‘VALUE nm_nd_row(int, VALUE*, VALUE)’:
../../../../ext/nmatrix/storage/yale/yale.cpp:1911:73: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘size_t {aka unsigned int}’ [-Wformat=]
     rb_raise(rb_eRangeError, "out of range (%lu >= %lu)", i, s->shape[0]);
                                                                         ^
../../../../ext/nmatrix/storage/yale/yale.cpp:1911:73: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 4 has type ‘size_t {aka unsigned int}’ [-Wformat=]
../../../../ext/nmatrix/storage/yale/yale.cpp: In function ‘VALUE nm_vector_set(int, VALUE*, VALUE)’:
../../../../ext/nmatrix/storage/yale/yale.cpp:1994:104: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘size_t {aka unsigned int}’ [-Wformat=]
     rb_raise(rb_eArgError, "lengths must match between j array (%lu) and value array (%lu)", len, vvlen);
                                                                                                        ^
../../../../ext/nmatrix/storage/yale/yale.cpp:1994:104: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 4 has type ‘size_t {aka unsigned int}’ [-Wformat=]
../../../../ext/nmatrix/storage/yale/yale.cpp: In instantiation of ‘char nm::yale_storage::vector_insert(YALE_STORAGE*, size_t, size_t*, void*, size_t, bool) [with DType = unsigned char; size_t = unsigned int]’:
../../../../ext/nmatrix/storage/yale/yale.cpp:1261:3:   required from here
../../../../ext/nmatrix/storage/yale/yale.cpp:574:127: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘size_t {aka unsigned int}’ [-Wformat=]
     rb_raise(rb_eArgError, "vector insert pos (%lu) is before beginning of ja (%lu); this should not happen", pos, s->shape[0]);
                                                                                                                               ^
../../../../ext/nmatrix/storage/yale/yale.cpp:574:127: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 4 has type ‘size_t {aka unsigned int}’ [-Wformat=]
../../../../ext/nmatrix/storage/yale/yale.cpp: In instantiation of ‘char nm::yale_storage::vector_insert(YALE_STORAGE*, size_t, size_t*, void*, size_t, bool) [with DType = signed char; size_t = unsigned int]’:
../../../../ext/nmatrix/storage/yale/yale.cpp:1261:3:   required from here
../../../../ext/nmatrix/storage/yale/yale.cpp:574:127: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘size_t {aka unsigned int}’ [-Wformat=]
../../../../ext/nmatrix/storage/yale/yale.cpp:574:127: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 4 has type ‘size_t {aka unsigned int}’ [-Wformat=]
../../../../ext/nmatrix/storage/yale/yale.cpp: In instantiation of ‘char nm::yale_storage::vector_insert(YALE_STORAGE*, size_t, size_t*, void*, size_t, bool) [with DType = short int; size_t = unsigned int]’:
../../../../ext/nmatrix/storage/yale/yale.cpp:1261:3:   required from here
../../../../ext/nmatrix/storage/yale/yale.cpp:574:127: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘size_t {aka unsigned int}’ [-Wformat=]
../../../../ext/nmatrix/storage/yale/yale.cpp:574:127: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 4 has type ‘size_t {aka unsigned int}’ [-Wformat=]

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,6 +8,8 @@ rvm:
   - 2.0
   - 2.1
   - 2.2
+  - 2.3.0
+  - ruby-head
 before_install: ./travis.sh before_install
 script: ./travis.sh script
 
@@ -23,6 +25,14 @@ matrix:
     - rvm: 2.2
       env: USE_REF=1 # Installs reference implementations of LAPACK and BLAS, builds and tests nmatrix, nmatrix-lapacke
     - rvm: 2.2
+      env: NO_EXTERNAL_LIB=1 # No external libraries installed, only nmatrix
+    - rvm: 2.3.0
+      env: USE_REF=1 # Installs reference implementations of LAPACK and BLAS, builds and tests nmatrix, nmatrix-lapacke
+    - rvm: 2.3.0
+      env: NO_EXTERNAL_LIB=1 # No external libraries installed, only nmatrix
+    - rvm: ruby-head
+      env: USE_REF=1 # Installs reference implementations of LAPACK and BLAS, builds and tests nmatrix, nmatrix-lapacke
+    - rvm: ruby-head
       env: NO_EXTERNAL_LIB=1 # No external libraries installed, only nmatrix
 notifications:
   irc: "chat.freenode.net#sciruby"

--- a/ext/nmatrix/extconf.rb
+++ b/ext/nmatrix/extconf.rb
@@ -164,6 +164,11 @@ CONFIG['warnflags'].gsub!('-Wshorten-64-to-32', '') # doesn't work except in Mac
 CONFIG['warnflags'].gsub!('-Wdeclaration-after-statement', '')
 CONFIG['warnflags'].gsub!('-Wimplicit-function-declaration', '')
 
+have_func("rb_array_const_ptr", "ruby.h")
+have_macro("FIX_CONST_VALUE_PTR", "ruby.h")
+have_macro("RARRAY_CONST_PTR", "ruby.h")
+have_macro("RARRAY_AREF", "ruby.h")
+
 create_conf_h("nmatrix_config.h")
 create_makefile("nmatrix")
 

--- a/ext/nmatrix/math.cpp
+++ b/ext/nmatrix/math.cpp
@@ -1002,7 +1002,7 @@ static VALUE nm_clapack_getrs(VALUE self, VALUE order, VALUE trans, VALUE n, VAL
   } else {
     ipiv_ = NM_ALLOCA_N(int, RARRAY_LEN(ipiv));
     for (int index = 0; index < RARRAY_LEN(ipiv); ++index) {
-      ipiv_[index] = FIX2INT( RARRAY_PTR(ipiv)[index] );
+      ipiv_[index] = FIX2INT( RARRAY_AREF(ipiv, index) );
     }
   }
 
@@ -1054,7 +1054,7 @@ static VALUE nm_clapack_laswp(VALUE self, VALUE n, VALUE a, VALUE lda, VALUE k1,
   } else {
     ipiv_ = NM_ALLOCA_N(int, RARRAY_LEN(ipiv));
     for (int index = 0; index < RARRAY_LEN(ipiv); ++index) {
-      ipiv_[index] = FIX2INT( RARRAY_PTR(ipiv)[index] );
+      ipiv_[index] = FIX2INT( RARRAY_AREF(ipiv, index) );
     }
   }
 

--- a/ext/nmatrix/nmatrix.h
+++ b/ext/nmatrix/nmatrix.h
@@ -57,6 +57,33 @@
   #include "nm_memory.h"
 #endif
 
+#ifndef FIX_CONST_VALUE_PTR
+# if defined(__fcc__) || defined(__fcc_version) || \
+    defined(__FCC__) || defined(__FCC_VERSION)
+/* workaround for old version of Fujitsu C Compiler (fcc) */
+#  define FIX_CONST_VALUE_PTR(x) ((const VALUE *)(x))
+# else
+#  define FIX_CONST_VALUE_PTR(x) (x)
+# endif
+#endif
+
+#ifndef HAVE_RB_ARRAY_CONST_PTR
+static inline const VALUE *
+rb_array_const_ptr(VALUE a)
+{
+  return FIX_CONST_VALUE_PTR((RBASIC(a)->flags & RARRAY_EMBED_FLAG) ?
+    RARRAY(a)->as.ary : RARRAY(a)->as.heap.ptr);
+}
+#endif
+
+#ifndef RARRAY_CONST_PTR
+# define RARRAY_CONST_PTR(a) rb_array_const_ptr(a)
+#endif
+
+#ifndef RARRAY_AREF
+# define RARRAY_AREF(a, i) (RARRAY_CONST_PTR(a)[i])
+#endif
+
 /*
  * Macros
  */

--- a/ext/nmatrix/ruby_nmatrix.c
+++ b/ext/nmatrix/ruby_nmatrix.c
@@ -2636,7 +2636,7 @@ nm::dtype_t nm_dtype_guess(VALUE v) {
      * TODO: Look at entire array for most specific type.
      */
 
-    return nm_dtype_guess(RARRAY_PTR(v)[0]);
+    return nm_dtype_guess(RARRAY_AREF(v, 0));
 
   default:
     RB_P(v);
@@ -2798,7 +2798,7 @@ static void* interpret_initial_value(VALUE arg, nm::dtype_t dtype) {
     init_val = NM_ALLOC_N(char, DTYPE_SIZES[dtype] * RARRAY_LEN(arg));
     NM_CHECK_ALLOC(init_val);
     for (index = 0; index < RARRAY_LEN(arg); ++index) {
-      rubyval_to_cval(RARRAY_PTR(arg)[index], dtype, (char*)init_val + (index * DTYPE_SIZES[dtype]));
+      rubyval_to_cval(RARRAY_AREF(arg, index), dtype, (char*)init_val + (index * DTYPE_SIZES[dtype]));
     }
 
   } else {
@@ -2825,7 +2825,7 @@ static size_t* interpret_shape(VALUE arg, size_t* dim) {
     shape = NM_ALLOC_N(size_t, *dim);
 
     for (size_t index = 0; index < *dim; ++index) {
-      shape[index] = FIX2UINT( RARRAY_PTR(arg)[index] );
+      shape[index] = FIX2UINT( RARRAY_AREF(arg, index) );
     }
 
   } else if (FIXNUM_P(arg)) {

--- a/ext/nmatrix_atlas/extconf.rb
+++ b/ext/nmatrix_atlas/extconf.rb
@@ -240,6 +240,11 @@ CONFIG['warnflags'].gsub!('-Wshorten-64-to-32', '') # doesn't work except in Mac
 CONFIG['warnflags'].gsub!('-Wdeclaration-after-statement', '')
 CONFIG['warnflags'].gsub!('-Wimplicit-function-declaration', '')
 
+have_func("rb_array_const_ptr", "ruby.h")
+have_macro("FIX_CONST_VALUE_PTR", "ruby.h")
+have_macro("RARRAY_CONST_PTR", "ruby.h")
+have_macro("RARRAY_AREF", "ruby.h")
+
 create_conf_h("nmatrix_atlas_config.h")
 create_makefile("nmatrix_atlas")
 

--- a/ext/nmatrix_atlas/math_atlas.cpp
+++ b/ext/nmatrix_atlas/math_atlas.cpp
@@ -1019,7 +1019,7 @@ static VALUE nm_atlas_clapack_getrs(VALUE self, VALUE order, VALUE trans, VALUE 
   } else {
     ipiv_ = NM_ALLOCA_N(int, RARRAY_LEN(ipiv));
     for (int index = 0; index < RARRAY_LEN(ipiv); ++index) {
-      ipiv_[index] = FIX2INT( RARRAY_PTR(ipiv)[index] );
+      ipiv_[index] = FIX2INT( RARRAY_AREF(ipiv, index) );
     }
   }
 
@@ -1104,7 +1104,7 @@ static VALUE nm_atlas_clapack_getri(VALUE self, VALUE order, VALUE n, VALUE a, V
   } else {
     ipiv_ = NM_ALLOCA_N(int, RARRAY_LEN(ipiv));
     for (int index = 0; index < RARRAY_LEN(ipiv); ++index) {
-      ipiv_[index] = FIX2INT( RARRAY_PTR(ipiv)[index] );
+      ipiv_[index] = FIX2INT( RARRAY_AREF(ipiv, index) );
     }
   }
 
@@ -1191,7 +1191,7 @@ static VALUE nm_atlas_clapack_laswp(VALUE self, VALUE n, VALUE a, VALUE lda, VAL
   } else {
     ipiv_ = NM_ALLOCA_N(int, RARRAY_LEN(ipiv));
     for (int index = 0; index < RARRAY_LEN(ipiv); ++index) {
-      ipiv_[index] = FIX2INT( RARRAY_PTR(ipiv)[index] );
+      ipiv_[index] = FIX2INT( RARRAY_AREF(ipiv, index) );
     }
   }
 

--- a/ext/nmatrix_lapacke/extconf.rb
+++ b/ext/nmatrix_lapacke/extconf.rb
@@ -194,6 +194,11 @@ CONFIG['warnflags'].gsub!('-Wshorten-64-to-32', '') # doesn't work except in Mac
 CONFIG['warnflags'].gsub!('-Wdeclaration-after-statement', '')
 CONFIG['warnflags'].gsub!('-Wimplicit-function-declaration', '')
 
+have_func("rb_array_const_ptr", "ruby.h")
+have_macro("FIX_CONST_VALUE_PTR", "ruby.h")
+have_macro("RARRAY_CONST_PTR", "ruby.h")
+have_macro("RARRAY_AREF", "ruby.h")
+
 create_conf_h("nmatrix_lapacke_config.h")
 create_makefile("nmatrix_lapacke")
 

--- a/ext/nmatrix_lapacke/math_lapacke.cpp
+++ b/ext/nmatrix_lapacke/math_lapacke.cpp
@@ -652,7 +652,7 @@ static VALUE nm_lapacke_lapacke_getri(VALUE self, VALUE order, VALUE n, VALUE a,
   } else {
     ipiv_ = NM_ALLOCA_N(int, RARRAY_LEN(ipiv));
     for (int index = 0; index < RARRAY_LEN(ipiv); ++index) {
-      ipiv_[index] = FIX2INT( RARRAY_PTR(ipiv)[index] );
+      ipiv_[index] = FIX2INT( RARRAY_AREF(ipiv, index) );
     }
   }
 
@@ -744,7 +744,7 @@ static VALUE nm_lapacke_lapacke_getrs(VALUE self, VALUE order, VALUE trans, VALU
   } else {
     ipiv_ = NM_ALLOCA_N(int, RARRAY_LEN(ipiv));
     for (int index = 0; index < RARRAY_LEN(ipiv); ++index) {
-      ipiv_[index] = FIX2INT( RARRAY_PTR(ipiv)[index] );
+      ipiv_[index] = FIX2INT( RARRAY_AREF(ipiv, index) );
     }
   }
 

